### PR TITLE
Reintroduce old definition struct and parser for remote builds

### LINKED ISSUE #3194: Remote builds are broken
## Version of Singularity:

master

### Expected behavior

`$ singularity build --remote test.sif docker://ubuntu:latest` works

### Actual behavior

```
$ singularity build --remote test.sif docker://ubuntu:latest
...
FATAL:   While performing build: failed to post request to remote build service: Failed to decode request: json: cannot unmarshal object into Go struct field ImageScripts.help of type string (400 Bad Request)
```

This is caused by the changes in the `definition` struct from #3156. The remote builder is using a different version of the struct so it fails when trying to unmarshall the json of the new struct. To fix this the `remotebuilder` package should use the older parser and `definition` struct

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

This adds a legacy version of the `definition` struct and parser to `/pkg/build`.
This was done in order to maintain compatibility with the remote builder since recent changes to the definition struct has broken the builders ability to unmarshall the definition json.

**This fixes or addresses the following GitHub issues:**

- Fixes #3194 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: pkg/build/legacy/definition.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline pkg/build/legacy/definition.go]:
```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```

File: pkg/build/legacy/parser/deffile.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]:
```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```

File: pkg/build/legacy/parser/deffile.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+	// trim % prefix on section name
+	line = strings.TrimLeft(line, "%")
+	lineSplit := strings.SplitN(strings.ToLower(line), " ", 2)
+
+	return lineSplit[0]
+}
+
+// parseTokenSection splits the token into maximum 2 strings separated by a newline,
+// and then inserts the section into the sections map
+//
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]: Stray `//` in program.

```suggestion
```

File: pkg/build/legacy/parser/deffile.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+func parseTokenSection(tok string, sections map[string]string) error {
+	split := strings.SplitN(tok, "\n", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("Section %v: Could not be split into section name and body", split[0])
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]:
```suggestion
		return fmt.Errorf("section %v: could not be split into section name and body", split[0])
```

File: pkg/build/legacy/parser/deffile.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+	}
+
+	key := getSectionName(split[0])
+	if appSections[key] {
+		sectionSplit := strings.SplitN(strings.TrimLeft(split[0], "%"), " ", 3)
+		if len(sectionSplit) < 2 {
+			return fmt.Errorf("App Section %v: Could not be split into section name and app name", sectionSplit[0])
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]:
```suggestion
			return fmt.Errorf("app section %v: could not be split into section name and app name", sectionSplit[0])
```

File: pkg/build/legacy/parser/deffile.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+func doSections(s *bufio.Scanner, d *types.Definition) error {
+	sectionsMap := make(map[string]string)
+
+	tok := strings.TrimSpace(s.Text())
+
+	// skip initial token parsing if it is empty after trimming whitespace
+	if tok != "" {
+		//check if first thing parsed is a header/comment or just a section
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]: I would add a space here for consistency:

```suggestion
		// Check if first thing parsed is a header/comment or just a section
```

File: pkg/build/legacy/parser/deffile.go [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+		if tok[0] != '%' {
+			if err := doHeader(tok, d); err != nil {
+				return fmt.Errorf("failed to parse DefFile header: %v", err)
+			}
+		} else {
+			//this is a section
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]: Here too:

```suggestion
			// This is a section
```

File: pkg/build/legacy/parser/deffile.go [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+			if err := parseTokenSection(tok, sectionsMap); err != nil {
+				return err
+			}
+		}
+	}
+
+	//parse remaining sections while scanner can advance
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]: And here:

```suggestion
	// Parse remaining sections while scanner can advance
```

File: pkg/build/legacy/parser/deffile.go [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+	return
+}
+
+// ParseDefinitionFile receives a reader from a definition file
+// and parse it into a Definition struct or return error if
+// the definition file has a bad section.
+func ParseDefinitionFile(r io.Reader) (d types.Definition, err error) {
+	d.Raw, err = ioutil.ReadAll(r)
+	if err != nil {
+		return d, fmt.Errorf("While attempting to read in definition: %v", err)
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile.go]:
```suggestion
		return d, fmt.Errorf("while attempting to read in definition: %v", err)
```

File: pkg/build/legacy/parser/deffile_test.go [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+//	if ident != ident_str || content != content_str {
+//		t.Fatal("splitToken returned bad values")
+//	}
+//
+//	str = "%%" + ident_str
+//	ident, content = splitToken(str)
+//	if ident != ident_str || content != "" {
+//		t.Fatal("splitToken returned bad values")
+//	}
+//}
```
Comment (WestleyK) [inline pkg/build/legacy/parser/deffile_test.go:91]: Extra code?

**Review by WestleyK — CHANGES_REQUESTED** [review]
Looks good, I have a couple suggestions :+1:

File: pkg/build/legacy/parser/deffile_test.go (hunk 10 above)
Comment (ikaneshiro) [reply pkg/build/legacy/parser/deffile_test.go:91]: I'm not sure. @gvallee was this intended to be commented out when it was originally committed in #3084?

File: pkg/build/legacy/parser/deffile.go (hunk 5 above)
Comment (WestleyK) [reply pkg/build/legacy/parser/deffile.go]: How did CCI/Travis pass with the caps in the error message? I thought it was a requirement that theres no caps in error messages?...

Comment (ikaneshiro) [reply pkg/build/legacy/parser/deffile.go]: I don't think it's a hard requirement, just good practice.

File: pkg/build/legacy/remote.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline pkg/build/legacy/remote.go]:
```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```
Or:
```suggestion
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

File: internal/pkg/build/remotebuilder/remotebuilder.go [hunk 12]
```diff
@@ -3,6 +3,12 @@
 // LICENSE.md file distributed with the sources of this project regarding your
 // rights to use or distribute this software.
 
+//
+// NOTE: This package uses a different version of the definition struct and
+// definition parser than the rest of the image build system in order to maintain
+// compatibility with the remote builder.
+//
+
```
Comment (WestleyK) [inline internal/pkg/build/remotebuilder/remotebuilder.go:11]: The copyright in this file should be:

```
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

**Review by WestleyK — COMMENTED** [review]
Couple small suggestions. After that, looks good to me :+1:

**Review by WestleyK — APPROVED** [review]
Looks good to me :+1:  :shipit:

**Pull request merged**